chalk-solve/src/wf.rs

```diff
@@ -41,63 +42,70 @@ pub struct WfSolver<'db, I: Interner> {
     solver_choice: SolverChoice,
 }
 
-/// A trait for retrieving all types appearing in some Chalk construction.
-///
-/// FIXME: why is this not a `Folder`?
-trait FoldInputTypes: HasInterner {
-    fn fold(&self, interner: &Self::Interner, accumulator: &mut Vec<Ty<Self::Interner>>);
+struct InputTypeCollector<'i, I: Interner> {
+    types: Vec<Ty<I>>,
+    interner: &'i I,
 }
 
-impl<T: FoldInputTypes> FoldInputTypes for [T] {
-    fn fold(&self, interner: &T::Interner, accumulator: &mut Vec<Ty<T::Interner>>) {
-        for f in self {
-            f.fold(interner, accumulator);
+impl<'i, I: Interner> InputTypeCollector<'i, I> {
+    fn new(interner: &'i I) -> Self {
+        Self {
+            types: Vec::new(),
+            interner,
         }
     }
 }
 
-impl<T: FoldInputTypes> FoldInputTypes for Vec<T> {
-    fn fold(&self, interner: &T::Interner, accumulator: &mut Vec<Ty<T::Interner>>) {
-        for f in self {
-            f.fold(interner, accumulator);
-        }
+impl<'i, 't, I: Interner> Visitor<'i, I> for InputTypeCollector<'i, I> {
+    type Result = ();
+
+    fn as_dyn(&mut self) -> &mut dyn Visitor<'i, I, Result = Self::Result> {
+        self
     }
-}
 
-impl<I: Interner> FoldInputTypes for Parameter<I> {
-    fn fold(&self, interner: &I, accumulator: &mut Vec<Ty<I>>) {
-        if let ParameterKind::Ty(ty) = self.data(interner) {
-            ty.fold(interner, accumulator)
-        }
+    fn interner(&self) -> &'i I {
+        self.interner
     }
-}
 
-impl<I: Interner> FoldInputTypes for Substitution<I> {
-    fn fold(&self, interner: &I, accumulator: &mut Vec<Ty<I>>) {
-        self.parameters(interner).fold(interner, accumulator)
+    fn visit_where_clause(&mut self, where_clause: &WhereClause<I>, outer_binder: DebruijnIndex) {
+        match where_clause {
+            WhereClause::AliasEq(alias_eq) => {
+                TyData::Alias(alias_eq.alias.clone())
+                    .intern(self.interner)
+                    .visit_with(self, outer_binder);
+                alias_eq.ty.visit_with(self, outer_binder);
```

Comment:
I think this could be 

```rust
if let WhereClause::AliasEq(alias_eq) = self {
TyData::Alias(alias_eq.alias.clone())
                    .intern(self.interner)
                    .visit_with(self, outer_binder);
}

where_clause.super_visit_with(self, outer_binder);
```

Although maybe this code is better, actually, in the sense of being more robust against future changes that might add new cases we should consider.

Reply:
(In case unclear, just leave as is I think)